distsqlrun: add a draining phase to all the distsql producers

andreimatei opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13570)
<!-- Reviewable:end -->

andreimatei committed:
distsqlrun: rename RowReceiver.Close to ProducerDone

andreimatei committed:
distsqlrun: change the communication protocol

... between producers and consumers. Also add a draining phase to all
the distsql producers.

Errors now flow inline with data, as opposed to trailing it. Metadata
has been added, which also flows inline with regular data rows.

Draining:
The idea is that a consumer (of rows) can tell the producer that it
doesn't need any more rows to satisfy the query, but will continue to
forward metadata. The consumer at this point is supposed to only produce
said metadata before informing the consumer that it's done.
This is intended to be used primarily by queries with LIMITs, where the
consumer controls the end of communication.

The metadata is initially going to consist on info about the ranges used
(range descriptors and lease information), to be used by the gateway to
figure out if it made bad planning decisions and update its caches. This
commit doesn't actually hook up the producing of metadata at the source
(TableReader) or the consumption at the gateway; these will come next.

Most of everything here is about routing metadata. Everything that used
to carry a row now carries either a row or metadata. Thus, the
producer-consumer protocol changes slightly, so see changes in
distsqlrun/base.go.
The on-the-wire protocol also changes and we've moved to bi-directional
gRPC streams (in order for the "please drain" signal from a consumer to
be able to flow between nodes) - see the proto changes. These
bi-directional streams will surely come in handy for other types of
communications that will need to flow from "consumer" to "producer" -
e.g. row requests on the PK side for a JoinReader.

Besides the protocol changes, the rest of the diff is about different
processors and components dealing with draining. I haven't introduced
any particular abstraction to deal with this; instead there are some
commonly used utility functions. I'm open to various ideas people might
have about potentially not burdening all the processors with draining
notions.

RaduBerinde commented:
Can/should `RowChannelMsg.Error` be part of the `ProducerMetadata`? I think it makes sense conceptually (the producer is telling us it hit an error) and should make things a bit simpler.

Can we make `NextRow()` return both a row and a `*ProducerMetadata` so the callers can do `row, meta := NextRow()` instead of having `row.Row()` everywhere? Similarly, `PushRow` should take two arguments instead of the caller having to construct a `RowOrMetadata`. Note: if you disagree, we should change the "row" variable names because they are very deceiving - `row.Empty()` does not mean what it reads.

Also maybe rename to `Next` and `Push` since we're not necessarily getting/pushing a row.

I suggest prototyping any changes here - instead of going through all processors, refactor just the noop processor so we can see how it looks. You can send me a pointer to the commit and we can talk offline.

andreimatei commented:
Great suggestions. I should have put errors inline with data from the get-go, but I chickened out. Errors are now part of metadata, and carried inline with the rest of the flow as opposed to trailing and generally being treated separately. I think this is an improvement.

Please give another read to `base.go`, `inbound/outbox.go`, the proto API, and take it like there.
I've changed everything but the tests.

cc @asubiotto @knz in case their lives is missing some excitement.

---

Review status: 0 of 43 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-Kd3e6mbAzJL23K_elmu:b-tgrxrc)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Definitely much nicer this way

---

Review status: 0 of 43 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/base.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6uXgeYUTZ-pNBjstv:-Kd6uXggVyiVm_F1Wnzj:blaqg0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L113)):*
> ```Go
> 	// point on (which means, for example, that they're not required to
> 	// automatically ask every producer to drain, in case there's multiple
> 	// producers). Therefor, consumers need to be aware that some rows might have
> ```

Therefore

---

*[pkg/sql/distsqlrun/base.go, line 272 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6umS5vJPzK3cKD96X:-Kd6umS5vJPzK3cKD96Y:boanqtb) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L272)):*
> ```Go
> 	// Only one of these fields will be set.
> 	Row  sqlbase.EncDatumRow
> 	Meta ProducerMetadata
> ```

Would it be better for this to be a pointer? `ProducerMetadata` is about 40(?) bytes which we will be copying and sending along with each row. We don't expect metadata-heavy flows to be the norm so an extra allocation when we have metadata should be fine.

If `NextRow() `returns `*ProducerMetadata` then we could just check for `nil` instead of `Empty()`

I am not sure if this would matter much. I have a benchmark (I used it to decide whether to use one channel for both rows and err or two channels). I added this case, but my CPU is very inconsistent with benchmarks, maybe you can run it - just clone this repo: https://github.com/RaduBerinde/playground/blob/master/chan_test/main_test.go

---

*[pkg/sql/distsqlrun/data.proto, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6blMThG1IB6UPBmfp:-Kd6blMThG1IB6UPBmfq:b33vdcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/data.proto#L177)):*
> ```protobuf
>     repeated roachpb.RangeInfo range_info = 1 [(gogoproto.nullable) = false];
>   }
>   oneof value {
> ```

can we use `gogoproto.onlyone`? Just so we don't mix these (the processors use onlyone)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-Kd70YlhP8X00jS2VNQk:bhjvict)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/distsqlrun/base.go, line 272 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6umS5vJPzK3cKD96X:-Kd7qOTRCtcSSAB2WZRi:b-192ofr) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L272)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Would it be better for this to be a pointer? `ProducerMetadata` is about 40(?) bytes which we will be copying and sending along with each row. We don't expect metadata-heavy flows to be the norm so an extra allocation when we have metadata should be fine.

If `NextRow() `returns `*ProducerMetadata` then we could just check for `nil` instead of `Empty()`

I am not sure if this would matter much. I have a benchmark (I used it to decide whether to use one channel for both rows and err or two channels). I added this case, but my CPU is very inconsistent with benchmarks, maybe you can run it - just clone this repo: https://github.com/RaduBerinde/playground/blob/master/chan_test/main_test.go
</blockquote></details>

Never mind, figured it out, there was a bug in the benchmarks.

Cost per msg with meta :
```
Chan3-4   69.2ns ± 2%
Chan3a-4   156ns ± 7%   // this variant also listens on a ticker channel
```

Cost per msg with *meta (set to `nil`):
```
Chan4-4   66.7ns ± 7%
Chan4a-4   152ns ± 5%   // this variant also listens on a ticker channel
```

Seems that it won't matter, feel free to leave it as non-pointer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Looked over most of the core stuff (base, inbound, outbox, protos). Awesome work!

---

Review status: 0 of 43 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/api.proto, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI2S001G6IEHXzNjwk:-KdI2S01GNFmubmTbYnq:b-hoam1) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/api.proto#L85)):*
> ```protobuf
>   // FlowStream is used to push a stream of messages that is part of a flow. The
>   // first message will have a StreamHeader which identifies the flow and the
>   // stream (mailbox).
> ```

add to the comment here hinting why the return is also a stream

---

*[pkg/sql/distsqlrun/base.go, line 55 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHxpqDhRBv8X214c2n:-KdHxpqDhRBv8X214c2o:b-c3kbbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L55)):*
> ```Go
> 	// NeedMoreRows indicates that the consumer is still expecting more rows.
> 	NeedMoreRows ConsumerStatus = iota
> 	// DrainRequested indicates that the consumer will not process any more data
> ```

very nice descriptions

---

*[pkg/sql/distsqlrun/base.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHy8xMaa7l4nyyn3oN:-KdHy8xMaa7l4nyyn3oO:byfasht) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L101)):*
> ```Go
> 	Types() []sqlbase.ColumnType
> 
> 	// NextRow returns the next record that a producer has pushed into this
> ```

[nit] Could call this just `Next` (and `PushRow` -> `Push`)

---

*[pkg/sql/distsqlrun/base.go, line 148 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHyjfSGlcH3QkLx5ue:-KdHyjfSGlcH3QkLx5uf:buprevh) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L148)):*
> ```Go
> //
> // It is OK to call DrainAndForwardMetadata() multiple times concurrently,
> // passing the same RowReceiver (as RowReceiver.PushRow() is guaranteed to be
> ```

[nit] "multiple times concurrenlty on the same dst" is a bit clearer

---

*[pkg/sql/distsqlrun/base.go, line 160 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHyyIW1gb7cwa1HjS1:-KdHyyIW1gb7cwa1HjS2:b-dflpqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L160)):*
> ```Go
> 			return
> 		}
> 		if meta.Empty() {
> ```

[nit] we can move `if row == nil { return }` here, we are checking for meta.Empty twice

---

*[pkg/sql/distsqlrun/base.go, line 171 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzGoew8Hh5LaGkIwY:-KdHzGofJNOkC678ecJI:b-9oink8) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L171)):*
> ```Go
> 		case NeedMoreRows:
> 			log.Fatal(ctx, "NeedMoreRows while draining")
> 		case DrainRequested:
> ```

[nit] move this at the end

---

*[pkg/sql/distsqlrun/base.go, line 179 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzYhvT2QFP2HxtqzB:-KdHzYhvT2QFP2HxtqzC:brgkq6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L179)):*
> ```Go
> }
> 
> // DrainAndForwardMetadataFinal is a version of DrainAndForwardMetadata that
> ```

maybe `DrainAndClose` since the semantics are different from DrainAndForwardMetadata

---

*[pkg/sql/distsqlrun/base.go, line 181 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzVnC3cSRYupEc4b-:-KdHzVnC3cSRYupEc4b0:baja19g) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L181)):*
> ```Go
> // DrainAndForwardMetadataFinal is a version of DrainAndForwardMetadata that
> // drains multiple sources. These sources are assumed to be the only producers
> // left for dst, so dst is closed once they're all exhauster (this is different
> ```

exhausted

---

*[pkg/sql/distsqlrun/base.go, line 194 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzj5FuabARVBTVSdC:-KdHzj5FuabARVBTVSdD:bbnpo3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L194)):*
> ```Go
> 	ctx context.Context, dst RowReceiver, cause error, srcs ...RowSource,
> ) {
> 	if cause == nil {
> ```

did you mean `!= nil`?

---

*[pkg/sql/distsqlrun/base.go, line 197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI-Qx2iQUp4LALmKMf:-KdI-Qx2iQUp4LALmKMg:bza60bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L197)):*
> ```Go
> 		consumerStatus := dst.PushRow(nil /* row */, ProducerMetadata{Err: cause})
> 		if consumerStatus == ConsumerClosed {
> 			dst.ProducerDone()
> ```

don't we need to `ConsumerClosed()` the sources?

---

*[pkg/sql/distsqlrun/base.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI0zplDHt8aorptc9J:-KdI0zplDHt8aorptc9K:b-8capno) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L213)):*
> ```Go
> }
> 
> // NoMetadataRowSource is a wrapper on top of a RowSource that automatically
> ```

This is really cool!

---

*[pkg/sql/distsqlrun/base.go, line 359 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI-skA7-brHdchQu-R:-KdI-skA7-brHdchQu-S:b-kj9z0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L359)):*
> ```Go
> // ConsumerDone is part of the RowSource interface.
> func (rc *RowChannel) ConsumerDone() {
> 	atomic.StoreUint32((*uint32)(&rc.consumerStatus), uint32(DrainRequested))
> ```

maybe first check if the existing value is `ConsumerClosed` and panic in that case? (to detect invalid usage)

---

*[pkg/sql/distsqlrun/base.go, line 416 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI08RUYPRL48He3HaP:-KdI08RUYPRL48He3HaQ:bb7od0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L416)):*
> ```Go
> // ConsumerDone is part of the RowSource interface.
> func (mrc *MultiplexedRowChannel) ConsumerDone() {
> 	atomic.StoreUint32(
> ```

same

---

*[pkg/sql/distsqlrun/base.go, line 434 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI0JzCDdEplSka4pdO:-KdI0JzCDdEplSka4pdP:b-jrwhoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L434)):*
> ```Go
> }
> 
> type BufferedRecord struct {
> ```

will need a comment

---

*[pkg/sql/distsqlrun/base.go, line 441 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI0RDcHDid05E02LoF:-KdI0RDcHDid05E02LoG:b1sbiql) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L441)):*
> ```Go
> // RowBuffer is an implementation of RowReceiver that buffers (accumulates)
> // results in memory, as well as an implementation of RowSource that returns
> // rows from a row buffer.
> ```

records from a record buffer

---

*[pkg/sql/distsqlrun/inbound.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI5QLpm1prCDc6CtFf:-KdI5QLpm1prCDc6CtFg:b3xpqnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/inbound.go#L39)):*
> ```Go
> 	// as the last record that the producer gets.
> 
> 	defer func() {
> ```

could move the big loop to another function and just call that first instead of using `defer` (easier to read that way, and we don't need named returns)

---

*[pkg/sql/distsqlrun/inbound.go, line 66 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI5nl5n2o8dFzZPzEv:-KdI5nl5n2o8dFzZPzEw:b-z2pmu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/inbound.go#L66)):*
> ```Go
> 				if err != io.EOF {
> 					// Communication error.
> 					finalErr = errors.Wrap(
> ```

:thumbsup: on the wrapping and tagging

---

*[pkg/sql/distsqlrun/inbound.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI6BIFkG9gD4LwdfGC:-KdI6BIFkG9gD4LwdfGD:br5heeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/inbound.go#L130)):*
> ```Go
> // producer that it doesn't need any more rows and the producer should drain. A
> // signal is sent on stream to the producer to ask it to send metadata.
> func sendDrainSignalToProducer(ctx context.Context, stream DistSQL_FlowStreamServer) error {
> ```

ToStreamProducer (the name is a bit too generic, this only applies to streams)

---

*[pkg/sql/distsqlrun/outbox.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI6eJ6Sc_aaX2tmCI1:-KdI6eJ6Sc_aaX2tmCI2:bn6upey) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/outbox.go#L259)):*
> ```Go
> 	ch := make(chan drainSignal, 1)
> 
> 	go func() {
> ```

kind of unfortunate that we set up a goroutine for every outbound stream. Add a TODO here to investigate optimizing this (OTOH maybe we could set this up after a delay so we avoid it if the query is fast enough?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-KdI7dkTP-heAxyI6QD6:b9wsw2f)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
distsqlrun: send stream header quickly

Before this patch, the header was sent together with the first batch or
rows. This was potentially bad, since the receiver really wants to
receive the header quickly, so it can finish the handshake and consider
the inbound stream connected (otherwise, after a timeout, it gives up on
waiting).
There was also a bug since the last commit causing us to never send the
header on a stream that wasn't sending any rows.

andreimatei committed:
util/log: export makeMessage

So other packages can create messages tagged with a context's tags,
similar to log messages.

andreimatei commented:
Fixed a few snags. If you're so inclined, please look at the diff from before.
Still have to write tests for errors and draining :S

---

Review status: 0 of 46 files reviewed at latest revision, 22 unresolved discussions.

---

*[pkg/sql/distsqlrun/api.proto, line 85 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI2S001G6IEHXzNjwk:-KdWhLI3svSG3JeEkhJz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/api.proto#L85)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

add to the comment here hinting why the return is also a stream
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHy8xMaa7l4nyyn3oN:-KdmeOSl1EXKGTa5snYO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L101)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Could call this just `Next` (and `PushRow` -> `Push`)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 113 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6uXgeYUTZ-pNBjstv:-KdWhf05Zd901bEX2tgj:bpqq6ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L113)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Therefore
</blockquote></details>

ah, there is a different between "therefore" and "therefore". TIL.

---

*[pkg/sql/distsqlrun/base.go, line 148 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHyjfSGlcH3QkLx5ue:-KdWi7xB3YqZGkehSB87:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L148)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] "multiple times concurrenlty on the same dst" is a bit clearer
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 160 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHyyIW1gb7cwa1HjS1:-KdWibXWnSWNZX7SySGo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L160)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] we can move `if row == nil { return }` here, we are checking for meta.Empty twice
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 171 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzGoew8Hh5LaGkIwY:-KdWihA8bnaN99HXBQwQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L171)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] move this at the end
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 179 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzYhvT2QFP2HxtqzB:-KdmeZkY9oVNuDn7rWR8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L179)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

maybe `DrainAndClose` since the semantics are different from DrainAndForwardMetadata
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 181 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzVnC3cSRYupEc4b-:-KdWiyiMxOcfefrJxolu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L181)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

exhausted
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 194 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdHzj5FuabARVBTVSdC:-KdWj0HxqLJiAg2HTeEU:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L194)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

did you mean `!= nil`?
</blockquote></details>

done

---

*[pkg/sql/distsqlrun/base.go, line 197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI-Qx2iQUp4LALmKMf:-KdWju8xomNrgNJaO8uZ:b-fq2kur) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L197)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

don't we need to `ConsumerClosed()` the sources?
</blockquote></details>

yeah, I guess we do. I've removed this code and left `DrainAndForwardMetadata()` below to always close the sources.

---

*[pkg/sql/distsqlrun/base.go, line 359 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI-skA7-brHdchQu-R:-KdWlGyZKt9rcaASLHjN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L359)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

maybe first check if the existing value is `ConsumerClosed` and panic in that case? (to detect invalid usage)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 416 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI08RUYPRL48He3HaP:-KdWljbozynloJ116NDK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L416)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 434 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI0JzCDdEplSka4pdO:-KdWlx4v3yy4xQ3jWPHl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L434)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

will need a comment
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/base.go, line 441 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI0RDcHDid05E02LoF:-KdWm61xcCxOhL-ISSli:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/base.go#L441)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

records from a record buffer
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/data.proto, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6blMThG1IB6UPBmfp:-KdWmA7wGsbCHHPSFOz9:bi15hwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/data.proto#L177)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

can we use `gogoproto.onlyone`? Just so we don't mix these (the processors use onlyone)
</blockquote></details>

I tried using `onlyone` first, but had some problems with it. `onlyone` is weird with `nullable = false` fields - it's documented to not support them, but in practice it appears to do support them unless those messages are also `onlyone`. It looks like in the current form of these protos we no longer have `onlyone` inside `onlyone` (I think `ProducerData` was also `onlyone` initially), but, still, I'd generally stick with `oneof`, as it's the official one.

I was also talking to @tamird about this the other day. The conclusion of that conversation was that we only want `onlyone` for really large unions, like the kv `Request` (and also the distsql processors), because there all the (many) generated structs and interfaces required by `oneof` really become a pain.

---

*[pkg/sql/distsqlrun/inbound.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI5QLpm1prCDc6CtFf:-KdX6d3ddzWtBgCdljP5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/inbound.go#L39)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

could move the big loop to another function and just call that first instead of using `defer` (easier to read that way, and we don't need named returns)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/inbound.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI6BIFkG9gD4LwdfGC:-KdXGDgDag5hbZtwTnlJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/inbound.go#L130)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

ToStreamProducer (the name is a bit too generic, this only applies to streams)
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/outbox.go, line 259 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KdI6eJ6Sc_aaX2tmCI1:-KdXGFBRhyXCr-psPpaX:bqenwut) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/outbox.go#L259)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

kind of unfortunate that we set up a goroutine for every outbound stream. Add a TODO here to investigate optimizing this (OTOH maybe we could set this up after a delay so we avoid it if the query is fast enough?)
</blockquote></details>

added a TODO... I don't really know what best to do, I hope to talk about it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-Kdmho984vBTTsdq5zt6:bje79dk)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 46 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/data.proto, line 177 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-Kd6blMThG1IB6UPBmfp:-KdmiXiCKLYYWNQY2OsO:b-unsjc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/aab4cc29f6b52352f053c1aa9b95a049ff093836/pkg/sql/distsqlrun/data.proto#L177)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I tried using `onlyone` first, but had some problems with it. `onlyone` is weird with `nullable = false` fields - it's documented to not support them, but in practice it appears to do support them unless those messages are also `onlyone`. It looks like in the current form of these protos we no longer have `onlyone` inside `onlyone` (I think `ProducerData` was also `onlyone` initially), but, still, I'd generally stick with `oneof`, as it's the official one.

I was also talking to @tamird about this the other day. The conclusion of that conversation was that we only want `onlyone` for really large unions, like the kv `Request` (and also the distsql processors), because there all the (many) generated structs and interfaces required by `oneof` really become a pain.
</blockquote></details>

This is not the reason for using `onlyone`; `onlyone` is useful in the presence of polymorphism, when a particular field is meant to be populated by one of many concrete types which are always assigned as an interface (e.g. proto.Request via SetValue).

In cases where you're assigning concrete types, `oneof` is always better, since it gives you better type safety. So: please prefer to use `oneof` whenever possible; proto.Request is a special snowflake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM

![x](https://media.giphy.com/media/dOJt6XZlQw8qQ/giphy.gif)

andreimatei commented:
added some tests. The last round of changes deserves another review. @arjunravinarayan, if you don't mind, please take a look.

---

Review status: 0 of 46 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-Kf9PaByXd4rsfxKjnQr:b-6jpvwm)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Great work, lots of comments which helps things a lot. Doing this review in phases and using the "mark files as reviewed" as the guide. Commencing phase 2 now, but sending out phase 1 review to keep this moving along.

---

Reviewed 1 of 45 files at r3, 3 of 17 files at r7, 13 of 43 files at r8.
Review status: 9 of 46 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/base.go, line 67 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfD3UV6AFQYj_HY4Ock:-KfD3UV6AFQYj_HY4Ocl:bdctuor) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/base.go#L67)):*
> ```Go
> // component. It can be an input synchronizer, a router, or a mailbox.
> type RowReceiver interface {
> 	// Push sends a record to the consumer of this RowReceiver. Exactly one of the
> ```

Exactly one? Also non-nil is a weird way to phrase it as `ProducerMetadata` is passed by value so when "nil" it's  actually an empty struct.

---

*[pkg/sql/distsqlrun/distinct.go, line 104 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDLxYfC_OPrQBHtBdA:-KfDLxYgo3Wr6nOD6s7s:b-90lvfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/distinct.go#L104)):*
> ```Go
> 		encoding, err = d.encode(scratch, row)
> 		if err != nil {
> 			cleanup()
> ```

So we just throw away errors now? That doesn't seem good. I think I'm missing something...

---

*[pkg/sql/distsqlrun/distinct.go, line 124 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDM3j2IK3SqGaKcKFS:-KfDM3j2IK3SqGaKcKFT:bwd0536) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/distinct.go#L124)):*
> ```Go
> 			d.seen[key] = struct{}{}
> 			if !emitHelper(ctx, &d.out, row, ProducerMetadata{}, d.input) {
> 				// No cleanup required; emitHelper() took care of it.
> ```

I don't like this pattern that `emitHelper` has introduced, this is dangerous, especially since if we screw this up, there's a `panic()` waiting for the poor fool who calls an accidental second cleanup after calling `emitHelper`.

---

*[pkg/sql/distsqlrun/processors.go, line 218 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDJbyKa6-fLI_uW7cR:-KfDJbyKa6-fLI_uW7cS:b-899xs1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/processors.go#L218)):*
> ```Go
> // rendering processing; the output has not been closed.
> //
> // Note: check out emitHelper() for a useful wrapper.
> ```

Can reword to "`emitHelper()` is a wrapper function that provides a gentler interface, rather than calling `emitRow()` directly.", but nbd.

---

*[pkg/sql/distsqlrun/routers.go, line 143 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDIInYqWUxdntPxejC:-KfDIInYqWUxdntPxejD:b7m7zyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/routers.go#L143)):*
> ```Go
> }
> 
> // fwdMetadata forwards a metadata record to the first stream that's still
> ```

why only the first stream? Apologies if this is documented elsewhere, but this is a large PR and hard to review, but why not all streams?

---

*[pkg/sql/distsqlrun/routers.go, line 154 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDI7YRQYIKgWve0zmH:-KfDI7YRQYIKgWve0zmI:blwgl9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/routers.go#L154)):*
> ```Go
> 			newStatus := rb.streams[i].Push(nil /*row*/, meta)
> 			rb.updateStreamState(i, newStatus)
> 			if newStatus != ConsumerClosed {
> ```

don't need `forwarded` at all, just do `if ... { return }`.

---

*[pkg/sql/distsqlrun/stream_merger.go, line 213 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDIkWoF9XqOez5hIGe:-KfDIkWoF9XqOez5hIGf:bbjy9sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/stream_merger.go#L213)):*
> ```Go
> }
> 
> // makeStreamMerger creates a streamMerger, joinging rows from leftSource with
> ```

`joinging`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-KfDInWHwupdCMkVZDzb:bfwlda5)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 45 files at r3, 30 of 43 files at r8, 9 of 9 files at r9.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/aggregator.go, line 145 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDN1sJndrK-Ki8qrzZ:-KfDN1sJndrK-Ki8qrz_:b-ctjah1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/aggregator.go#L145)):*
> ```Go
> 
> 	if err := ag.accumulateRows(ctx); err != nil {
> 		// We swallow the error here, it has already been forwarded to the output.
> ```

This is a weird pattern. If `accumulateRows` handles error propagation, then it should return a boolean. Then this comment won't be needed.

---

*[pkg/sql/distsqlrun/data.proto, line 167 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDQrPA3U34kJN3CvnP:-KfDQrPBt2eORDZMcKPP:b-g3vdfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9459eaadd7872f94a3dbc138200e8c725cf7d6e0/pkg/sql/distsqlrun/data.proto#L167)):*
> ```protobuf
>   // This field has to be populated on, or before, a ProducerMessage with data
>   // in it, and can only be populated once.
>   // TODO(andrei): It'd be nice if the typing information for streams would be
> ```

I agree, this whole pattern is weird, but it's good for now.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 126 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDO-1MzCNiHEwiGodx:-KfDO-1MzCNiHEwiGody:bfmzsb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/hashjoiner.go#L126)):*
> ```Go
> // entirely from the right stream with the encoding/group key generated using the
> // left equality columns. If a row is found to have a NULL in an equality column
> // (and thus will not match anything), it might be routed directly to the
> ```

"Might"? You should explain why/when this would happen.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-KfDQubSOXvKDPtGtR3L:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
TFYR!

---

Review status: 3 of 46 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/sql/distsqlrun/aggregator.go, line 145 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDN1sJndrK-Ki8qrzZ:-KfDt_J585xU6uzbO2Q3:bcvhu5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/aggregator.go#L145)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

This is a weird pattern. If `accumulateRows` handles error propagation, then it should return a boolean. Then this comment won't be needed.
</blockquote></details>

mmm the pattern of both forwarding the error and returning the error is also used at least in the mailbox, which can both forward an error on a stream and save the error to be returned by `RunSyncFlow`. When you have two servers involved, I think it can make sense to both forward the error and report it up the local stack.
I don't know what to say; I think returning the error is better than a success bool, and I think a comment here about the state of the inputs and/or output would still be needed here.

---

*[pkg/sql/distsqlrun/base.go, line 67 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfD3UV6AFQYj_HY4Ock:-KfDy5a4HoErcq6CiUms:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/base.go#L67)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Exactly one? Also non-nil is a weird way to phrase it as `ProducerMetadata` is passed by value so when "nil" it's  actually an empty struct.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/distinct.go, line 104 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDLxYfC_OPrQBHtBdA:-KfDzPOON7Z3NfFIEtv8:bixifk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/distinct.go#L104)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

So we just throw away errors now? That doesn't seem good. I think I'm missing something...
</blockquote></details>

oops, I think this just got lost in the shuffle. good catch.

---

*[pkg/sql/distsqlrun/distinct.go, line 124 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDM3j2IK3SqGaKcKFS:-KfE-PojlbkENYfB1Ryb:b58tw48) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/distinct.go#L124)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I don't like this pattern that `emitHelper` has introduced, this is dangerous, especially since if we screw this up, there's a `panic()` waiting for the poor fool who calls an accidental second cleanup after calling `emitHelper`.
</blockquote></details>

ack, but I'm not sure what would be better. I went back and forth on a couple of things around this when writing it. The difficulty is that, every time you push data, you have to deal with a multitude of situations - consumer closed, consumer wants draining, errors. And we push data to consumers all over the place, sometimes in multiple places in a single processor. And so if `emitHelper` wouldn't handle cleanup, we'd just end up with more complicated code everywhere else - which, if anything, would exacerbate the danger.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 126 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDO-1MzCNiHEwiGodx:-KfE1R8Db3E-Zz4ukInU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/hashjoiner.go#L126)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

"Might"? You should explain why/when this would happen.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/routers.go, line 143 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDIInYqWUxdntPxejC:-KfE1hniMn9coHP7odln:b6u4x57) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/routers.go#L143)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

why only the first stream? Apologies if this is documented elsewhere, but this is a large PR and hard to review, but why not all streams?
</blockquote></details>

The goal is to get metadata records to the gateway node. We only need them to travel through one path.
Blasting them everywhere can create duplicates later on.

---

*[pkg/sql/distsqlrun/routers.go, line 154 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDI7YRQYIKgWve0zmH:-KfE2GkDnhj71MeCWuCo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/routers.go#L154)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

don't need `forwarded` at all, just do `if ... { return }`.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/stream_merger.go, line 213 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-KfDIkWoF9XqOez5hIGe:-KfE2M1CLaBVcUXTmYtQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee0901e70330f925024c25632991d308db9ebc0b/pkg/sql/distsqlrun/stream_merger.go#L213)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

`joinging`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13570#-:-KfE3cQfAhul3U8bWHKK:b-814cj3)*
<!-- Sent from Reviewable.io -->
